### ISSUE TITLE
Add support for the MULTISET operator

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#3884]
The SQL standard specifies a very interesting ORDBMS feature: `ARRAY` and `MULTISET` which are essentially nested tables.

An example:

``` sql
SELECT
  first_name,
  last_name,
  MULTISET(
    SELECT * FROM book WHERE author_id = author.id
  ) books
FROM author
```

Multiset are a much more intuitive way to join data in a hierarchical manner, rather than joining them via `JOIN` clauses in the `FROM` clause. In particular, these nested tables are much more similar to the way object oriented design works, by nesting type hierarchies.

Databases that support `ARRAY`, `MULTISET`, or `SET`:

- BigQuery (`ARRAY`)
- Informix (`LIST` = `ARRAY`, `MULTISET`, `SET`)
- H2 (`ARRAY`)
- Oracle (`VARRAY`, `TABLE` are the types, and `MULTISET` is the operator)
- PostgreSQL `ARRAY`

Other dialects can emulate `ARRAY` using XML or JSON in:

- [x] Projections (`MULTISET` and `SET` could also be emulated)
- [x] Predicates (`MULTISET` and `SET` are harder to emulate)

We need a new `Settings.emulateMultiset` with the following options:

- [x] `NestedCollectionEmulation.DEFAULT` - Use a native implementation if possible, or the "most optimal emulation", otherwise
- [x] ~`NestedCollectionEmulation.ARRAY` - Force using the ARRAY emulation~ (postponed to the next release: https://github.com/jOOQ/jOOQ/issues/12021)
- [x] `NestedCollectionEmulation.XML` - Force using the XML emulation
- [x] `NestedCollectionEmulation.JSON` - Force using the JSON emulation
- [x] `NestedCollectionEmulation.JSONB` - Force using the JSONB emulation
- [x] `NestedCollectionEmulation.NATIVE` - Force using the native implementation

### API

**`ARRAY`**

We already have `DSL.array(...)` which produces a `Field<T[]>` type. While it is possible since jOOQ 3.15 (https://github.com/jOOQ/jOOQ/issues/11812) to project `Field<Record[N][]>` types using `array(select(row(T.A, T.B)))`, that explicit nesting of `ROW()` into an `ARRAY` type is cumbersome (despite that's what PG offers).

We want some form of `array(select(T.A, T.B))` producing `Field<Record2<AType, BType>[]>`, though it will not be possible to overload `DSL.array()` like that, so another name will be needed, e.g. `arrayX()` (a placeholder for now).

(Note, there exists a feature request to change `Field<T[]>` to `Field<List<T>>`: https://github.com/jOOQ/jOOQ/issues/12010)

**`MULTISET` and `SET`**

Similar to the previous `array()` API, new `multiset(Select<R>): Field<Result<R>>` and `set(Select<R>): Field<Result<R>>` API is added. This time, we won't use arrays as the result type, but much more useful `Result` types, which will be more helpful when mapping nested collections. A similar naming convention as with `arrayX()` will be needed and applied here, for consistency reasons.

### Rejected functionality

- There is a possibility of `MULTISET` usage in `SELECT` clauses being transformed into separate queries that allow for emulating this feature - at least for simple join conditions. 
  
  There are too many open questions in this area, and the SQL/XML or SQL/JSON emulation is too tempting. Let's focus on this, first.

### Future work

- See this issue https://github.com/jOOQ/jOOQ/issues/12021

**Comment by nakolkin** [comment]
Hi Lucas, this feature looks very promising! Wondering if there is ongoing work to actually implement it?  In the meantime, could you suggest any best practices to fetch/map object hierarchies with JOOQ? I've read this blog post https://blog.jooq.org/2014/04/11/java-8-friday-no-more-need-for-orms/ but maybe there are some updates in that field. Thanks!

**Comment by lukaseder** [comment]
No, there hasn't been any progress on this feature - more important features have been prioritised in the recent past.

We usually recommend using a third party mapper, such as simpleflatmapper by @arnaudroger
https://github.com/arnaudroger/SimpleFlatMapper

**Comment by lukaseder** [comment]
Picking this up again for jOOQ 3.15.

jOOQ 3.14 now supports the necessary underlying infrastructure to emulate this in dialects without native support, i.e. everywhere:

- CUBRID support is deprecated
- Informix might still be able to do it
- Oracle requires named table and object types, which we don't have (we could create them on the fly, but that raises new questions, besides, Oracle has complete standard SQL/XML and SQL/JSON support)

An example of how nested collections can be mapped to POJOs using SQL/JSON and a mapping library like Jackson/Gson has been given in a few Stack Overflow questions, e.g. https://stackoverflow.com/a/64314221/521799, or in this blog post:
https://blog.jooq.org/2020/10/09/nesting-collections-with-jooq-3-14s-sql-xml-or-sql-json-support/

For example, in order to map into this data structure:

```java
class Student {
  List<Teacher> teachers;
  List<Book> books;
}
```

We can write this SQL using SQL/JSON

```java
List<Student> students =
ctx.select(jsonObject(
     jsonEntry("name", STUDENT.NAME),
     jsonEntry("id", STUDENT.ID),
     jsonEntry("teachers", field(
       select(jsonArrayAgg(jsonObject(TEACHER.NAME, TEACHER.ID)))
       .from(TEACHER)
       .join(STUDENT_TEACHER).on(TEACHER.ID.eq(STUDENT_TEACHER.TEACHER_ID))
       .where(STUDENT_TEACHER.STUDENT_ID.eq(STUDENT.ID))
     )),
     jsonEntry("books", field(
       select(jsonArrayAgg(jsonObject(BOOK.NAME, BOOK.ID)))
       .from(BOOK)
       .join(STUDENT_BOOK).on(BOOK.ID.eq(STUDENT_BOOK.BOOK_ID))
       .where(STUDENT_BOOK.STUDENT_ID.eq(STUDENT.ID))
     ))
   ))
   .from(STUDENT)
   .fetchInto(Student.class);
```

Using `MULTISET`, the following should be possible:

```java
List<Student> students =
ctx.select(
     STUDENT.NAME,
     STUDENT.ID,
     multiset(
       select(TEACHER.NAME, TEACHER.ID)
       .from(TEACHER)
       .join(STUDENT_TEACHER).on(TEACHER.ID.eq(STUDENT_TEACHER.TEACHER_ID))
       .where(STUDENT_TEACHER.STUDENT_ID.eq(STUDENT.ID))
     ).as("teachers"),
     multiset(
       select(BOOK.NAME, BOOK.ID)
       .from(BOOK)
       .join(STUDENT_BOOK).on(BOOK.ID.eq(STUDENT_BOOK.BOOK_ID))
       .where(STUDENT_BOOK.STUDENT_ID.eq(STUDENT.ID))
     ).as("books")
   )
   .from(STUDENT)
   .fetchInto(Student.class);
```

I.e. almost the same, except that:

- the `jsonXYZ()` boilerplate is removed
- there is now type information for the nested collections available. The resulting type (if not mapping into a pojo) will be: 
    ```java
    Result<Record4<String, Long, Result<Record2<String, Long>>, Result<Record2<String, Long>>>>
    ```

**Comment by JayAhn2** [comment]
@lukaseder Is the multiselect syntax planned to be released in 3.15?
I love this syntax lol 👍 
nice job! 💯

**Comment by lukaseder** [comment]
@JayAhn2 No promises 😉

**Comment by lukaseder** [comment]
OK, so it seems we now have enough internal infrastructure to pull off an experimental implementation. A quick draft shows that these work on MySQL using JSON, and on PostgreSQL using JSONB:

```sql
// Find all authors who have exactly books with ID = 1 and ID = 2
int count = ctx.fetchValue(
    selectCount()
   .from(AUTHOR)
   .where(
       multiset(
           select(BOOK.ID).from(BOOK).where(BOOK.AUTHOR_ID.eq(AUTHOR.ID))
       )
       .eq(multiset(
           select(val(1).as(BOOK.ID))
           .union(
           select(val(2).as(BOOK.ID)))
       ))
   )
);

// Get all authors and their book ID/TITLE pairs
Result<Record2<Integer, Result<Record2<Integer, String>>>> r2 =
ctx.select(AUTHOR.ID, multiset(
            select(BOOK.ID, BOOK.TITLE)
            .from(BOOK)
            .where(BOOK.AUTHOR_ID.eq(AUTHOR.ID))
        ))
        .from(AUTHOR)
        .orderBy(AUTHOR.ID)
        .fetch();
```

More sophisticated examples will follow, including combining the nesting with mapping to allow for type safe mapping into JDK 16 record type hierarchies

**Comment by lukaseder** [comment]
The XML emulation also works on PostgreSQL. Will look into the ARRAY/ROW emulation tomorrow, as well as all the other dialects, and many more tests, including:

- Aliasing
- Nesting of multisets
- Distinct and order by
- Unions and other set operations
- SQL Server `FOR JSON` and `FOR XML` emulations, which work a bit differently
- Oracle native support with types (probably to be postponed)

**Comment by lukaseder** [comment]
Instead of focusing only on `MULTISET`, we'll also enhance existing `ARRAY` support. An idea is to take inspiration from Infromix, which took ORDBMS features the furthest and has:

- A `LIST` type (similar to standard SQL `ARRAY`: Contents are ordered and accessible by index)
- A `MULTISET` type (similar to standard SQL `MULTISET`: Contents are not ordered and not accessible by index)
- A `SET` type (`MULTSET`, but `DISTINCT` applied to it: Contents are not ordered and not accessible by index)

I've updated the issue description accordingly.

**Comment by lukaseder** [comment]
Since we're going to be using standard jOOQ XML and JSON serialisation formats, we could offer configuration options for the various `JSONFormat.RecordFormat` (`ARRAY` and `OBJECT`) as well as `XMLFormat.RecordFormat` (`VALUE_ELEMENTS_WITH_FIELD_ATTRIBUTE`, `VALUE_ELEMENTS`, `COLUMN_NAME_ELEMENTS`)

**Comment by lukaseder** [comment]
To avoid any ambiguities with duplicate map keys (supported by JSON or XML but not necessarily in parsed Java representations, such as `Map`), we'll default to `ARRAY` and `VALUE_ELEMENTS` representations.

**Comment by lukaseder** [comment]
Future work is being tracked here: https://github.com/jOOQ/jOOQ/issues/12021. Specifically, the emulation using `ARRAY` types will be postponed.

**Comment by lukaseder** [comment]
Implemented in jOOQ 3.15.0

**Comment by lukaseder** [comment]
For now, `multiset()` should return a `SelectField<Result<R>>`, not a `Field<Result<R>>` type, to prevent using `MULTISET` in predicates until the `LIST` / `MULTISET` / `SET` semantics is implemented correctly. This will prevent users from using something unsupported accidentally, getting things that "work" sometimes, and later on running into regressions.

**Comment by lukaseder** [comment]
I feel the predicates bit has to be implemented correctly for 3.15 - otherwise, users will start depending on the non-deterministic behaviour, and when we fix this, we'll create breaking changes. Will do via a separate issue: https://github.com/jOOQ/jOOQ/issues/12030
